lib/open_feature/sdk/configuration.rb

```diff
@@ -13,13 +13,26 @@ module SDK
     class Configuration
       extend Forwardable
 
-      attr_accessor :context, :provider, :hooks
+      attr_accessor :context, :hooks
+      attr_reader :provider
 
       def_delegator :@provider, :metadata
 
       def initialize
         @hooks = []
       end
+
+      # When switching providers, there are a few lifecycle methods that need to be taken care of.
+      #   1. If a provider is already set, we need to call `shutdown` on it.
+      #   2. On the new provider, call `init`.
+      #   3. Finally, set the internal provider to the new provider
+      def provider=(provider)
+        @provider.shutdown if !@provider.nil? && @provider.respond_to?(:shutdown)
+
+        provider.init if provider.respond_to?(:init)
+
+        @provider = provider
+      end
```

Comment:
I'm sorry I must confess, I am not experienced with Ruby.

It looks like the provider setter is blocking (correct me if I'm wrong) which is good. Most SDKs implement both and blocking and non blocking version, but I would recommend you just implement the blocking one (meaning that setting the provider doesn't return until the providers `init` has completed or failed (in which case an exception should be thrown).

The shutdown doesn't need to block however, and an error should just be logged if a provider fails to shut down.

Reply:
Unfortunately, async paradigms aren't very common in Ruby. We could probably use the [async gem](https://github.com/socketry/async) but then we'd be introducing a dependency that might be unexpected for those pulling the SDK in. Is having shutdown be non-blocking important? Maybe we could spin up an issue on the repo and discuss all the places where allowing for async behavior would useful?

Reply:
We don't need async behavior in the technical sense.

What we need is for a method for setting the provider that "waits" (blocks or waits asynchronously) for the `init` of that provider to complete or fail before it returns and the next line executes. As long as we have that, it's no problem.

Reply:
Got it! So we will block until init finishes before returning to fulfill that requirement. Does that answer the question, or am I missing what you're asking?

Reply:
Nope, you got it. If that's how it works, all set!